packages/next/src/client/components/navigation-devtools.ts

```diff
@@ -0,0 +1,179 @@
+import type { FlightRouterState } from '../../shared/lib/app-router-types'
+import type { Params } from '../../server/request/params'
+import {
+  createDevToolsInstrumentedPromise,
+  type InstrumentedPromise,
+  type NavigationPromises,
+} from '../../shared/lib/hooks-client-context.shared-runtime'
+import {
+  computeSelectedLayoutSegment,
+  getSelectedLayoutSegmentPath,
+} from '../../shared/lib/segment'
+import { ReadonlyURLSearchParams } from './navigation.react-server'
+
+/**
+ * Promises are cached by tree to ensure stability across suspense retries.
+ */
+type LayoutSegmentPromisesCache = {
+  selectedLayoutSegmentPromises: Map<string, InstrumentedPromise<string | null>>
+  selectedLayoutSegmentsPromises: Map<string, InstrumentedPromise<string[]>>
+}
+
+const layoutSegmentPromisesCache = new WeakMap<
+  FlightRouterState,
+  LayoutSegmentPromisesCache
+>()
+
+/**
+ * Creates instrumented promises for layout segment hooks at a given tree level.
+ * This is dev-only code for React Suspense DevTools instrumentation.
+ */
+export function createLayoutSegmentPromises(
+  tree: FlightRouterState
+): LayoutSegmentPromisesCache | null {
+  if (process.env.NODE_ENV === 'production') {
+    return null
+  }
+
+  // Check if we already have cached promises for this tree
+  const cached = layoutSegmentPromisesCache.get(tree)
+  if (cached) {
+    return cached
+  }
+
+  // Create new promises and cache them
+  const segmentPromises = new Map<string, InstrumentedPromise<string | null>>()
+  const segmentsPromises = new Map<string, InstrumentedPromise<string[]>>()
+
+  const parallelRoutes = tree[1]
+  for (const parallelRouteKey of Object.keys(parallelRoutes)) {
+    const segments = getSelectedLayoutSegmentPath(tree, parallelRouteKey)
+
+    // Use the shared logic to compute the segment value
+    const segment = computeSelectedLayoutSegment(segments, parallelRouteKey)
+
+    segmentPromises.set(
+      parallelRouteKey,
+      createDevToolsInstrumentedPromise('useSelectedLayoutSegment', segment)
+    )
+    segmentsPromises.set(
+      parallelRouteKey,
+      createDevToolsInstrumentedPromise('useSelectedLayoutSegments', segments)
+    )
+  }
+
+  const result = {
```

Comment:
```suggestion
  const result: LayoutSegmentPromisesCache = {
```
because [TypeScript would not complain if you add more properties here than actually needed](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAogdsATiKBeKBvKAzA9rgLigGckBLOAcygF8AoAY1zlKgYEMGALadOCAO5QAsuzAAeUogqUANLATIAfAAoAlHUbNWERSnRY8hKAHJ2xACYn5AI3aIiARlvsAXkQBM89gBsfAVTgAV2IICyIAZlpGTh4AOlDgFRNidit5XSQQDTogA).